axelrod/strategies/punisher.py

```diff
@@ -143,3 +143,42 @@ def strategy(self, opponent: Player) -> Action:
             return D
         else:
             return C
+
+
+class TrickyLevelPunisher(Player):
+    """
+    A player starts by cooperating however, after 10 rounds
+    will defect if at any point the number of defections
+    by an opponent is greater than 20%.
+
+    Names:
+
+    - Tricky Level Punisher: [Eckhart2015]_
+    """
+
+    name = 'Tricky Level Punisher'
+    classifier = {
+        'memory_depth': float('inf'), # Long Memory
+        'stochastic': False,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def strategy(self, opponent: Player) -> Action:
+        if len(opponent.history) < 10:
+            return C
+        elif (len(opponent.history) - opponent.cooperations) / len(opponent.history) > 0.2:
+            return D
+        elif len(opponent.history) < 50:
```

Comment:
Looking at the strategy at https://github.com/jecki/CoopSim/blob/master/Strategies.py#L1022 this does not look like it's implemented correctly. The ratio checks are done after checking the number of turns.

So I believe this should look something like:

```python
if len(opponent.history) < 50:
    if (len(opponent.history) - opponent.cooperations) / len(opponent.history) > 0.2:
        return C
    return D
```

and similarly for the rest.

Essentially the strategy always checks the cooperation difference ratio against some level and that level changes depending on the number of turns.

Also (very minor) to avoid `len` calculations you can use:

```
opponent.defections / (opponent.defections + opponent.cooperations)
```

instead of:

```
(len(opponent.history) - opponent.cooperations) / len(opponent.history)
```

(I know you've done it your way because that's how it's done in `LevelPunisher`)